Reduce memory footprint in simulation runs.

the-x-at opened the pull request:
A complete simulation run is broken down into manageable chunks of simulations, 1000 at the moment. Data obtained in this way are added incrementally to dedicated data structures. Histograms and densities in the result objects are affected by this change, and will appear slightly different. This fixes issue #22. Needless to say that `R CMD check` finishes without complaints.

Tests corrected: Kinship sum test, Kinship group test.

Test not corrected: Familial incidence rate test.

Test not affected by this issue: Genealogical index of familiality test.

I have repeated the familial aggregation runs presented in the memory consumption table in issue #22. For documentation purposes, these runs are command line calls using the Institute for Biomedicine (EURAC) internal Gitlab `chris-famagg` package,

    maxmem time /home/shared/bioinf/R/bin/Rscript-4.0-BioC3.12 src/R/famaggize.R --skipkin --trait bu02-up-x0lp17 --method ks --plotter ks2paint --pedformat png --dataset c13000 --nsim N

where `N` is the number of simulation runs given in the following table:

_N_ | Time (s) | Max. mem (KB)| Time (s, fix)| Max. mem (KB, fix)
---------:|---------:|-------------:|------------:|-------------------:
1,000 |      61  |     642,848  |          68 |           593,224
2,000 |      92  |     564,812  |         102 |           594,156
4,000 |     132  |     619,124  |         138 |           625,572
8,000 |     236  |     731,824  |         255 |           656,464
16,000 |     437  |     900,644  |         463 |           585,188
32,000 |     799  |   1,368,572  |         725 |           645,312
64,000 |   1,400  |   2,558,352  |       1,553 |           610,064
128,000 |   3,088  |   5,054,656  |       3,001 |           624,660
256,000 |   5,794  |   8,911,296  |       6,109 |           647,648
512,000 |  12,445  |  15,579,764  |      11,814 |           638,508
1,024,000 |  24,803  |  34,090,076  |          NA |                NA

The first three columns are as in issue #22, the last two columns contain values measured with the memory fix presented here. Runtime is barely affected, the fluctuations are due to the workload of the time of the experiment. It appears there is a bit of an overhead introduced, expressed in runtime increments below 11% with a remarkable outlier for _N_ = 512,000 (-6%). We have refrained from running a million simulations, simply because the trend is so evident and to save time, resources, energy, and the planet.

the-x-at committed:
Reduce memory footprint in simulation runs.

A complete simulation run is broken down into manageable chunks of
simulations, 1000 at the moment. Data obtained in this way are added
incrementally to dedicated data structures. Histograms and densities in
the result objects are affected by this change, and will appear slightly
different. This fixes issue #22.

Tests corrected: Kinship sum test, Kinship group test.

Test not corrected: Familial incidence rate test.

Test not affected by this issue: Genealogical index of familiality test.

jorainer commented on R/Methods-FAKinGroupResults.R:
```diff
@@ -217,85 +217,100 @@ setMethod("runSimulation", "FAKinGroupResults", function(object, nsim=50000,
+    ## compute incremental tables and densities. A set of tables is ultimately
+    ## transformed to a histogram.
+
+    nr.inc.sim <- 1000
```
I would suggest to use here `nr.inc.sim <- min(1000, nr.sim.todo)` - to cover also the possibility `nsim < nr.sim.todo`.

the-x-at replied:
This is covered, the `if` statement in line 233 and the assignment in line 223 takes care of that. It's assigned before any simulation is done. But the `min` solution is more elegant. Will take it.

jorainer replied:
Thanks. I've also seen that - after writing this comment :)

jorainer commented on R/Methods-FAKinGroupResults.R:
```diff
@@ -217,85 +217,100 @@ setMethod("runSimulation", "FAKinGroupResults", function(object, nsim=50000,
+            })
+        }
+
+        BigRes <- list()
```
Since we know the length of `BigRes` beforehand we could already create it with `BigRes <- vector("list", nr.inc.sim)`. We need than also to change line 290 below (next comment).

jorainer commented on R/Methods-FAKinGroupResults.R:
```diff
@@ -217,85 +217,100 @@ setMethod("runSimulation", "FAKinGroupResults", function(object, nsim=50000,
+            }))
+            ExpRatio <- ExpRatio + ExpRes[, 1]
+            ExpKin   <- ExpKin   + ExpRes[, 2]
+            BigRes <- append(BigRes, list(ExpRes[, 3]))
```
If we created the `BigRes` list already with the right size above we could here use simply: `BigRes[[i]] <- ExpRes[, 3]` - this *should* avoid costly copying of the whole object because a single element is added. I'm however not sure if `list(ExpRes[, 3])` or `ExpRes[, 3]` should be added - I think `ExpRes[, 3]` should suffice, but that depends also how you access the data from `BigRes` below.

jorainer commented on R/Methods-FAKinGroupResults.R:
```diff
@@ -217,85 +217,100 @@ setMethod("runSimulation", "FAKinGroupResults", function(object, nsim=50000,
+        ## BigRes is now a list of length nr.inc.sim, each element a vector
+        ## of length affectedKins.
+        ExpVals <- do.call(rbind, BigRes)
+        ExpVals <- split(t(ExpVals), f=names(affectedKins))
```
What exactly do you do here? Is this `split` doing the correct thing? Note that if `ExpVals` is a `matrix` (which can well be with the proposed changes above) `split` will consider it as a `numeric` vector instead. If you want to explicitely split by row (i.e. `f` groups rows in the `matrix`/`data.frame`) you should use `split.data.frame` instead of `split`.

the-x-at replied:
This is Jo-code. I just took the same stuff you have done into the loop. It splits properly and as wanted. Doesn't look nice, though. I also wasn't too happy with this weird construct.

jorainer replied:
you mean it still splits properly even after the proposed change above? jo-code is bad - has improved over the last years, but the old one was sometimes not efficient...

jorainer reviewed (changes requested):
Thanks for the PR @the-x-at !
Some minor notes/comments that we should address/discuss before merging.

the-x-at committed:
Simplify code in Kinship group test

Better initializations and access to various numeric data. Slightly
nicer way to terminate simulation loop.

the-x-at commented:
Improved the items discussed. Looks better now. Also de-jo-ified said code chunk.
There is some problem with the package version number, that's also why the commit action fails under Linux. Code is validated in `R-4.1.0` with Bioconductor 3.13 with `R CMD check`.

jorainer reviewed (approved):
Looks good to me!

jorainer committed:
Merge branch 'master' into memory-fix

jorainer committed:
ci: update to BioC 3.14

jorainer commented:
I also updated the unit tests and bumped the version - will merge after the unit tests are done

jorainer commented:
Looks like Windows tests fails because binaries for BioC 3.14 are offline.
